sonarts-core/src/runner/sonartsServer.ts

```diff
@@ -26,86 +26,86 @@ import { getIssues } from "./rules";
 import { FileCache, createService } from "./languageService";
 import { parseTsConfig } from "../utils/parser";
 
-const fileCache = new FileCache();
-// key is ts file path, value is corresponding tsconfig path
-const tsConfigCache: Map<string, string> = new Map();
-const servicesPerTsconfig: Map<string, ts.LanguageService> = new Map();
-
 const EMPTY_ANSWER: { issues: any[] } = { issues: [] };
+export class SonarTsServer {
+  readonly fileCache = new FileCache();
+  // key is ts file path, value is corresponding tsconfig path
+  readonly tsConfigCache: Map<string, string> = new Map();
+  readonly servicesPerTsconfig: Map<string, ts.LanguageService> = new Map();
 
-function logTypeScriptMetaInfo() {
-  const version = require("typescript/package.json").version;
-  const location = path.dirname(path.dirname(require.resolve("typescript")));
+  public start(port: number) {
+    logTypeScriptMetaInfo();
+    const client = net.createConnection(port, "localhost", () => {
+      console.log("SonarTS Server connected to " + port);
+      let accumulatedData = "";
+      client.on("data", data => {
+        accumulatedData += data;
+        try {
+          const request = JSON.parse(accumulatedData.toString());
+          accumulatedData = "";
+          const issues = this.handleAnalysisRequest(request);
+          client.write(JSON.stringify(issues));
+        } catch (e) {
+          // ignore
+        }
+      });
+    });
+  }
 
-  console.log(`Using typescript at [${location}], version ${version}`);
-}
+  private handleAnalysisRequest(request: AnalysisRequest): any {
+    const { file, rules, content } = request;
 
-export function start(port: number) {
-  logTypeScriptMetaInfo();
+    this.fileCache.newContent({ file, content });
 
-  const client = net.createConnection(port, "localhost", () => {
-    console.log("SonarTS Server connected to " + port);
-    let accumulatedData = "";
-    client.on("data", data => {
-      accumulatedData += data;
-      try {
-        const request = JSON.parse(accumulatedData.toString());
-        accumulatedData = "";
-        const issues = handleAnalysisRequest(request);
-        client.write(JSON.stringify(issues));
-      } catch (e) {
-        // ignore
+    let tsConfig;
+    if (this.tsConfigCache.has(file)) {
+      tsConfig = this.tsConfigCache.get(file)!;
```

Comment:
![Bug](https://sonarsource.github.io/sonar-github/bug.png 'Bug') Bug: Forbidden non null assertion ([typescript:S2966](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=typescript%3AS2966))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=sonarts&branch=issue-456&issues=AWEj3AvYgvt_XO9IEk83&open=AWEj3AvYgvt_XO9IEk83)

[ISSUE_KEY]: AWEj3AvYgvt_XO9IEk83

Reply:
This issue has been resolved by a user in SonarQube

[ISSUE_STATUS]: RESOLVED